File: app/src/main/java/com/health/openworkout/gui/session/SessionFragment.java
package com.health.openworkout.gui.session;

import android.app.AlertDialog;
import android.content.res.Configuration;
import android.os.Bundle;
import android.text.InputType;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.Animation;
import android.view.animation.AnimationUtils;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.navigation.Navigation;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.google.android.material.textview.MaterialTextView;
import com.health.openworkout.R;
import com.health.openworkout.core.OpenWorkout;
import com.health.openworkout.core.datatypes.TrainingPlan;
import com.health.openworkout.core.datatypes.WorkoutItem;
import com.health.openworkout.core.datatypes.WorkoutSession;
import com.health.openworkout.gui.datatypes.GenericAdapter;
import com.health.openworkout.gui.datatypes.GenericFragment;
import com.health.openworkout.gui.datatypes.GenericSettingsFragment;

import java.util.List;

public class SessionFragment extends GenericFragment {
    private RecyclerView sessionsView;

    private TrainingPlan trainingPlan;
    private List<WorkoutSession> workoutSessionList;

    private FloatingActionButton expandableButton;
    private FloatingActionButton addButton;
    private LinearLayout addLayout;
    private Animation animFabOpen, animFabClose, animFabClock, animFabAntiClock;
    private boolean isExpandable;

    private SessionsAdapter sessionsAdapter;

    public View onCreateView(@NonNull LayoutInflater inflater,
                             ViewGroup container, Bundle savedInstanceState) {
        View root = inflater.inflate(R.layout.fragment_session, container, false);

        sessionsView = root.findViewById(R.id.sessionsView);

        sessionsView.setHasFixedSize(true);
        sessionsView.setLayoutManager(new GridLayoutManager(getContext(), getNumberOfColumns()));

        isExpandable = false;

        expandableButton = root.findViewById(R.id.expandableButton);
        addButton = root.findViewById(R.id.addButton);
        addLayout = root.findViewById(R.id.addLayout);

        animFabClose = AnimationUtils.loadAnimation(getContext(), R.anim.fab_close);
        animFabOpen = AnimationUtils.loadAnimation(getContext(), R.anim.fab_open);
        animFabClock = AnimationUtils.loadAnimation(getContext(), R.anim.fab_rotate_clock);
        animFabAntiClock = AnimationUtils.loadAnimation(getContext(), R.anim.fab_rotate_anticlock);

        expandableButton.setOnClickListener(v -> {
            if (isExpandable) {
                addLayout.setVisibility(View.GONE);
                addLayout.startAnimation(animFabClose);
                expandableButton.startAnimation(animFabAntiClock);
                isExpandable = false;
            } else {
                addLayout.setVisibility(View.VISIBLE);
                addLayout.startAnimation(animFabOpen);
                expandableButton.startAnimation(animFabClock);
                isExpandable = true;
            }
        });

        addButton.setOnClickListener(v -> {
            AlertDialog.Builder alert = new AlertDialog.Builder(getContext());
            alert.setTitle(getString(R.string.label_input_create_days));
            final EditText input = new EditText(getContext());
            input.setText("3", MaterialTextView.BufferType.EDITABLE);
            input.setInputType(InputType.TYPE_CLASS_NUMBER);
            input.setRawInputType(Configuration.KEYBOARD_12KEY);
            alert.setView(input);
            alert.setPositiveButton(getString(R.string.label_ok), (dialog, whichButton) -> {
                if (!input.getText().toString().isEmpty()) {
                    int startNr = trainingPlan.getWorkoutSessions().size() + 1;
                    int offsetNr = Integer.valueOf(input.getText().toString());

                    for (int nr=startNr; nr < (startNr + offsetNr); nr++) {
                        WorkoutSession workoutSession = new WorkoutSession();
                        workoutSession.name = String.format(getString(R.string.day_unit), nr);
                        workoutSession.trainingPlanId = trainingPlan.getTrainingPlanId();
                        workoutSession.orderNr = nr;
                        trainingPlan.addWorkoutSession(workoutSession);
                        OpenWorkout.getInstance().insertWorkoutSession(workoutSession);
                        getAdapter().notifyItemInserted(nr);
                        sessionsView.scrollToPosition(startNr);
                    }

                    loadFromDatabase();
                }
            });
            alert.setNegativeButton(getString(R.string.label_cancel), (dialog, whichButton) -> {
                // empty
            });
            alert.show();
        });

        loadFromDatabase();

        return root;
    }

    @Override
    protected String getTitle() {
        return trainingPlan.getName();
    }

    @Override
    protected GenericAdapter getAdapter() {
        return sessionsAdapter;
    }

    @Override
    protected RecyclerView getRecyclerView() {
        return sessionsView;
    }

    @Override
    protected List getItemList() {
        return workoutSessionList;
    }

    @Override
    protected void onSelectCallback(int position) {
        WorkoutSession workoutSession = workoutSessionList.get(position);

        SessionFragmentDirections.ActionSessionFragmentToWorkoutFragment action = SessionFragmentDirections.actionSessionFragmentToWorkoutFragment();
        action.setTitle(workoutSession.name);
        action.setSessionWorkoutId(workoutSession.workoutSessionId);
        Navigation.findNavController(getActivity(), R.id.nav_host_fragment).navigate(action);
    }

    @Override
    protected void onEditCallback(int position) {
        WorkoutSession workoutSession = workoutSessionList.get(position);

        SessionFragmentDirections.ActionSessionsFragmentToSessionSettingsFragment action = SessionFragmentDirections.actionSessionsFragmentToSessionSettingsFragment();
        action.setWorkoutSessionId(workoutSession.workoutSessionId);
        action.setMode(GenericSettingsFragment.SETTING_MODE.EDIT);
        action.setTitle(getString(R.string.label_edit));
        Navigation.findNavController(getActivity(), R.id.nav_host_fragment).navigate(action);
    }

    @Override
    protected void onDeleteCallback(int position) {
        OpenWorkout.getInstance().deleteWorkoutSession(workoutSessionList.get(position));
        Toast.makeText(getContext(), String.format(getString(R.string.label_delete_toast), workoutSessionList.get(position).name), Toast.LENGTH_SHORT).show();
        getItemList().remove(position);
    }

    @Override
    protected void onDuplicateCallback(int position) {
        WorkoutSession origWorkoutSession = workoutSessionList.get(position);
        WorkoutSession duplicateWorkoutSession = origWorkoutSession.clone();

        duplicateWorkoutSession.workoutSessionId = 0;
        workoutSessionList.add(position, duplicateWorkoutSession);
        saveToDatabase();

        long workoutSessionId = OpenWorkout.getInstance().insertWorkoutSession(duplicateWorkoutSession);
        duplicateWorkoutSession.workoutSessionId = workoutSessionId;
    }

    @Override
    protected void onResetClick() {
        for (WorkoutSession workoutSession : workoutSessionList) {
            workoutSession.setFinished(false);

            for (WorkoutItem workoutItem : workoutSession.getWorkoutItems()) {
                workoutItem.setFinished(false);
                OpenWorkout.getInstance().updateWorkoutItem(workoutItem);
            }

            OpenWorkout.getInstance().updateWorkoutSession(workoutSession);
        }
    }

    @Override
    protected void loadFromDatabase() {
        long trainingPlanId = SessionFragmentArgs.fromBundle(getArguments()).getTrainingPlanId();
        trainingPlan = OpenWorkout.getInstance().getTrainingPlan(trainingPlanId);

        workoutSessionList = trainingPlan.getWorkoutSessions();

        sessionsAdapter = new SessionsAdapter(getContext(), workoutSessionList);
        sessionsAdapter.setMode(getMode());
        sessionsView.setAdapter(sessionsAdapter);
    }

    @Override
    protected void saveToDatabase() {
        for (int i=0; i<workoutSessionList.size(); i++) {
            workoutSessionList.get(i).orderNr = i;
            OpenWorkout.getInstance().updateWorkoutSession(workoutSessionList.get(i));
        }
    }

    private int getNumberOfColumns() {
        View view = View.inflate(getContext(), R.layout.item_session, null);
        view.measure(View.MeasureSpec.UNSPECIFIED, View.MeasureSpec.UNSPECIFIED);
        int width = view.getMeasuredWidth();
        int count = (getResources().getDisplayMetrics().widthPixels - sessionsView.getPaddingLeft() - sessionsView.getPaddingRight()) / width;
        int remaining = (getResources().getDisplayMetrics().widthPixels - sessionsView.getPaddingLeft() - sessionsView.getPaddingRight()) - width * count;
        if (remaining > width - 15)
            count++;
        return count;
    }
}

File: app/src/main/java/com/health/openworkout/gui/session/SessionFragment.kt
package com.health.openworkout.gui.session

import android.app.AlertDialog
import android.content.res.Configuration
import android.os.Bundle
import android.text.InputType
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.animation.Animation
import android.view.animation.AnimationUtils
import android.widget.EditText
import android.widget.LinearLayout
import android.widget.Toast

import androidx.navigation.fragment.findNavController
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView

import com.google.android.material.floatingactionbutton.FloatingActionButton
import com.health.openworkout.R
import com.health.openworkout.core.OpenWorkout
import com.health.openworkout.core.datatypes.TrainingPlan
import com.health.openworkout.core.datatypes.WorkoutSession
import com.health.openworkout.gui.datatypes.GenericFragment
import com.health.openworkout.gui.datatypes.GenericSettingsFragment

class SessionFragment : GenericFragment() {

    private lateinit var sessionsView: RecyclerView
    private lateinit var trainingPlan: TrainingPlan
    private lateinit var workoutSessionList: MutableList<WorkoutSession>

    private lateinit var expandableButton: FloatingActionButton
    private lateinit var addButton: FloatingActionButton
    private lateinit var addLayout: LinearLayout
    private lateinit var animFabOpen: Animation
    private lateinit var animFabClose: Animation
    private lateinit var animFabClock: Animation
    private lateinit var animFabAntiClock: Animation
    private var isExpandable = false

    private lateinit var sessionsAdapter: SessionsAdapter

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?
    ): View? {
        val root = inflater.inflate(R.layout.fragment_session, container, false)

        sessionsView = root.findViewById(R.id.sessionsView)
        sessionsView.setHasFixedSize(true)
        sessionsView.layoutManager = GridLayoutManager(context, getNumberOfColumns())

        expandableButton = root.findViewById(R.id.expandableButton)
        addButton = root.findViewById(R.id.addButton)
        addLayout = root.findViewById(R.id.addLayout)

        animFabClose = AnimationUtils.loadAnimation(context, R.anim.fab_close)
        animFabOpen = AnimationUtils.loadAnimation(context, R.anim.fab_open)
        animFabClock = AnimationUtils.loadAnimation(context, R.anim.fab_rotate_clock)
        animFabAntiClock = AnimationUtils.loadAnimation(context, R.anim.fab_rotate_anticlock)

        expandableButton.setOnClickListener {
            if (isExpandable) {
                addLayout.visibility = View.GONE
                addLayout.startAnimation(animFabClose)
                expandableButton.startAnimation(animFabAntiClock)
                isExpandable = false
            } else {
                addLayout.visibility = View.VISIBLE
                addLayout.startAnimation(animFabOpen)
                expandableButton.startAnimation(animFabClock)
                isExpandable = true
            }
        }

        addButton.setOnClickListener {
            showCreateDaysDialog()
        }

        loadFromDatabase()

        return root
    }

    private fun showCreateDaysDialog() {
        val alert = AlertDialog.Builder(requireContext())
        alert.setTitle(getString(R.string.label_input_create_days))
        val input = EditText(context).apply {
            setText("3")
            inputType = InputType.TYPE_CLASS_NUMBER
            setRawInputType(Configuration.KEYBOARD_12KEY)
        }
        alert.setView(input)
        alert.setPositiveButton(getString(R.string.label_ok)) { _, _ ->
            val inputText = input.text.toString()
            if (inputText.isNotEmpty()) {
                val startNr = trainingPlan.workoutSessions.size + 1
                val offsetNr = inputText.toInt()

                for (nr in startNr until (startNr + offsetNr)) {
                    val workoutSession = WorkoutSession().apply {
                        name = getString(R.string.day_unit, nr)
                        trainingPlanId = trainingPlan.trainingPlanId
                        orderNr = nr.toLong()
                    }
                    trainingPlan.addWorkoutSession(workoutSession)
                    OpenWorkout.getInstance().insertWorkoutSession(workoutSession)
                    getAdapter().notifyItemInserted(nr)
                    sessionsView.scrollToPosition(startNr)
                }

                loadFromDatabase()
            }
        }
        alert.setNegativeButton(getString(R.string.label_cancel), null)
        alert.show()
    }

    override fun getTitle(): String = trainingPlan.name

    override fun getAdapter(): SessionsAdapter = sessionsAdapter

    override fun getRecyclerView(): RecyclerView = sessionsView

    override fun getItemList(): MutableList<WorkoutSession> = workoutSessionList

    override fun onSelectCallback(position: Int) {
        val workoutSession = workoutSessionList[position]
        val action = SessionFragmentDirections.actionSessionFragmentToWorkoutFragment().apply {
            title = workoutSession.name.toString()
            sessionWorkoutId = workoutSession.workoutSessionId
        }
        findNavController().navigate(action)
    }

    override fun onEditCallback(position: Int) {
        val workoutSession = workoutSessionList[position]
        val action = SessionFragmentDirections.actionSessionsFragmentToSessionSettingsFragment().apply {
            workoutSessionId = workoutSession.workoutSessionId
            mode = GenericSettingsFragment.SETTING_MODE.EDIT
            title = getString(R.string.label_edit)
        }
        findNavController().navigate(action)
    }

    override fun onDeleteCallback(position: Int) {
        OpenWorkout.getInstance().deleteWorkoutSession(workoutSessionList[position])
        Toast.makeText(
            context,
            getString(R.string.label_delete_toast, workoutSessionList[position].name),
            Toast.LENGTH_SHORT
        ).show()
        getItemList().removeAt(position)
    }

    override fun onDuplicateCallback(position: Int) {
        val origWorkoutSession = workoutSessionList[position]
        val duplicateWorkoutSession = origWorkoutSession.clone().apply {
            workoutSessionId = 0
        }
        workoutSessionList.add(position, duplicateWorkoutSession)
        saveToDatabase()

        val workoutSessionId = OpenWorkout.getInstance().insertWorkoutSession(duplicateWorkoutSession)
        duplicateWorkoutSession.workoutSessionId = workoutSessionId
    }

    override fun onResetClick() {
        workoutSessionList.forEach { workoutSession ->
            workoutSession.isFinished = false
            workoutSession.getWorkoutItems().forEach { workoutItem ->
                workoutItem.setFinished(false)
                OpenWorkout.getInstance().updateWorkoutItem(workoutItem)
            }
            OpenWorkout.getInstance().updateWorkoutSession(workoutSession)
        }
    }

    override fun loadFromDatabase() {
        val trainingPlanId = SessionFragmentArgs.fromBundle(requireArguments()).trainingPlanId
        trainingPlan = OpenWorkout.getInstance().getTrainingPlan(trainingPlanId)
        workoutSessionList = trainingPlan.workoutSessions
        sessionsAdapter = SessionsAdapter(requireContext(), workoutSessionList).apply {
            mode = mode
        }
        sessionsView.adapter = sessionsAdapter
    }

    override fun saveToDatabase() {
        workoutSessionList.forEachIndexed { i, workoutSession ->
            workoutSession.orderNr = i.toLong()
            OpenWorkout.getInstance().updateWorkoutSession(workoutSession)
        }
    }

    private fun getNumberOfColumns(): Int {
        val view = View.inflate(context, R.layout.item_session, null)
        view.measure(View.MeasureSpec.UNSPECIFIED, View.MeasureSpec.UNSPECIFIED)
        val width = view.measuredWidth
        val screenWidth = resources.displayMetrics.widthPixels
        val count = (screenWidth - sessionsView.paddingLeft - sessionsView.paddingRight) / width
        val remaining = (screenWidth - sessionsView.paddingLeft - sessionsView.paddingRight) - width * count
        return if (remaining > width - 15) count + 1 else count
    }
}
